### PR TITLE
Threadless shardio

### PR DESCRIPTION
**Pull request by pmarks**
'sender pays' threading model - sending thread is used to compress & write data when the sort buffer fills up, no threads are created.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -263,264 +213,322 @@ where
...
-                disk_chunk_size,
-                writer,
-                to_writer_recv,
-                to_buffer_send,
-            );
-            swt.process()
-        });
+        assert!(disk_chunk_size >= 1);
+        assert!(item_buffer_size >= 1);
+        assert!(sender_buffer_size >= 1);
```
Comment (sreenathkrishnan) [inline src/lib.rs]: Should we also assert for the relative values? Like `assert!(item_buffer_size > sender_buffer_size)`?

File: src/lib.rs [hunk 2]
```diff
@@ -263,264 +213,322 @@ where
...
+            // outcome could be to process a full vec, retry, or be done.
+            match outcome {
+                Process(mut buf_to_process) => {
+                    // process the buffer & return it to the pool.
+                    self.process_buffer(&mut buf_to_process)?;
+                    self.return_buffer(buf_to_process);
+                    break;
+                },
+                Retry => {
+                    thread::yield_now();
```
Comment (sreenathkrishnan) [inline src/lib.rs]: What does this yield do? Isn't the thread going to attempt to acquire the mutex lock in the next iteration right away?

Comment (sreenathkrishnan) [reply src/lib.rs]: Ah got it, it's giving the thread back to the scheduler.

File: src/lib.rs (hunk 1 above)
Comment (pmarks) [reply src/lib.rs]: good idea, added.

File: src/lib.rs [hunk 3]
```diff
@@ -541,55 +552,139 @@ where
...
+/// file manager maintains the index data.
+struct ShardWriterInner<T, S>
+where
+    T: Send + Serialize,
+    S: SortKey<T>,
+    <S as SortKey<T>>::Key: Ord + Clone + Serialize,
+{
+    sender_buffer_size: usize,
+    state_machine: BufferStateMachine<T, SortAndWriteHandler<T, S>>,
+    closed: AtomicBool,
```
Comment (sreenathkrishnan) [inline src/lib.rs]: Doesn't look like this AtomicBool is set to true anywhere. Is it needed in addition to the `closed` in `BufferStateMachine`?

File: Cargo.toml [hunk 4]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "shardio"
-version = "0.5.5"
```
Comment (nlhepler) [inline Cargo.toml]: Version ... decrement?

File: src/lib.rs (hunk 2 above)
Comment (nlhepler) [reply src/lib.rs]: I think this communication should be arbitered by a condvar. Retry should be blocking on it while the Process thread issues to it.

File: src/lib.rs (hunk 3 above)
Comment (pmarks) [reply src/lib.rs]: removed.

File: src/lib.rs (hunk 2 above)
Comment (pmarks) [reply src/lib.rs]: @nlhepler going to take this as a todo for the future. Things seem to be behaving ok with the current code.

**Pull request merged**